trainer/python/penrose/util/conf_util.py

```diff
@@ -0,0 +1,208 @@
+# limitations under the License.
+
+import inspect
+
+from penrose.util import reflect_util
+from penrose.util.log_util import default_logger as logger
+from pyhocon import ConfigFactory, ConfigMissingException, ConfigTree
+
+config_tree_get_ori = ConfigTree.get
+
+
+def new_func(*args, **kwds):
+    default_in_kw = kwds.get("default", None)
+    default_in_args = args[2] if len(args) > 2 else None
+    default = default_in_kw or default_in_args
+    try:
+        return config_tree_get_ori(*args, **kwds)
+    except ConfigMissingException:
+        logger.info(f"key {args[1]}, def {default}")
+        return default
+
+
+ConfigTree.get = new_func
+
+
+class Singleton:
+    _instance = None
+
+    def __new__(cls, *args, **kargs):
+        if not isinstance(cls._instance, cls):
+            cls._instance = super().__new__(cls)
+        return cls._instance
+
+
+class Configuration(Singleton):
+    """wrapper get and set for new conf"""
+
+    def __init__(self, inner_dict=None):
+        object.__setattr__(
+            self, "inner_dict", dict() if inner_dict is None else inner_dict
+        )
+
+    def __contains__(self, key):
+        return key in self.inner_dict
+
+    def get(self, key, default=None):
+        # check whether the key exists before
+        val_before = self.inner_dict.get(key, None)
+        val = self.inner_dict.get(key, default)
+
+        logger.info(
+            f"get value from conf: the key is {key} and value is {val}"
+        )
+        logger.info(f"get value from conf: value type is {type(val)}")
+        if type(val) is dict:
+            # to avoid ConfigTree is a sub-class of dict
+            val = ConfigFactory.from_dict(val)
+            if val_before is not None:
+                self.inner_dict[key] = val
+        return val
+
+    def put(self, key, value):
+        self.inner_dict[key] = value
+
+    def to_dict(self):
+        return self.inner_dict
+
+    def from_dict(self, dt):
+        if not isinstance(dt, dict):
+            logger.warning(f"you should pass dict not {type(dt)}")
+            return
+        for k, v in dt.items():
+            self.set(k, v)
+
+    def clear(self):
+        self.inner_dict.clear()
+
+    def __getitem__(self, key):
+        return self.inner_dict.get(key, None)
+
+    def __setitem__(self, key, value):
+        self.inner_dict[key] = value
+
+    def __str__(self):
+        return str(self.inner_dict)
+
+    def __getattr__(self, key):
+        if key.startswith("__"):
+            return object.__getattribute__(self, key)
+        return self.inner_dict.get(key, None)
+
+    def __setattr__(self, key, value):
+        self.inner_dict.setdefault(key, value)
+
+    def __reduce__(self):
+        return (self.__class__, (self.inner_dict,))
+
+
+class ConfigurationManagerMeta(type):
+    _all_conf_by_name = {}  # type:dict
+    _final_attrs = {}  # type:dict
+    _override_order = []  # type:list
```

Comment:
`_override_order: List[value_type] = []`

Reply:
def run( a:int, b:int) -> int: c:List[str] = ["2"] return a + b run(3,3)
it can run.

But the following usage is not legal
class A: aa:List[str] = ["2"] a = A() print(a.aa)